File: Game-1-modular/systems/potion_system.py
"""
Potion Effect Executor System
Tag-driven potion effect application system
"""

from typing import Dict, Tuple, Any, TYPE_CHECKING
from data.models.materials import MaterialDefinition
from entities.components.buffs import ActiveBuff

if TYPE_CHECKING:
    from entities.character import Character


class PotionEffectExecutor:
    """
    Executes potion effects based on tags and parameters.

    Supported Tags:
    - healing: Restore HP (instant or over_time)
    - mana_restore: Restore mana (instant or over_time)
    - buff: Apply stat buffs (strength, defense, speed, etc.)
    - resistance: Elemental damage resistance (fire, ice, elemental)
    - utility: Tool/armor enhancements (efficiency, armor, weapon)

    Tag Modifiers:
    - instant: Effect applies immediately
    - over_time: Effect applies gradually
    - self: Targets the character using the potion
    """

    def __init__(self):
        """Initialize the potion effect executor"""
        pass

    def apply_potion_effect(
        self,
        character: 'Character',
        potion_def: MaterialDefinition,
        crafted_stats: Dict[str, Any] = None
    ) -> Tuple[bool, str]:
        """
        Apply potion effects to character based on tags and parameters.

        Args:
            character: Character to apply effects to
            potion_def: Potion material definition with effect_tags and effect_params
            crafted_stats: Optional crafting quality stats (potency, duration, quality)

        Returns:
            Tuple[bool, str]: (success, message)
        """
        if crafted_stats is None:
            crafted_stats = {}

        # Get quality multipliers
        potency = crafted_stats.get('potency', 100) / 100.0  # Effect strength multiplier
        duration_mult = crafted_stats.get('duration', 100) / 100.0  # Duration multiplier

        tags = potion_def.effect_tags
        params = potion_def.effect_params

        if not tags:
            return False, f"{potion_def.name} has no effect tags defined"

        messages = []
        success = False

        # Process healing effects
        if 'healing' in tags:
            heal_success, heal_msg = self._apply_healing(character, tags, params, potency, duration_mult)
            if heal_success:
                success = True
                messages.append(heal_msg)

        # Process mana restoration effects
        if 'mana_restore' in tags:
            mana_success, mana_msg = self._apply_mana_restore(character, tags, params, potency, duration_mult)
            if mana_success:
                success = True
                messages.append(mana_msg)

        # Process buff effects
        if 'buff' in tags:
            buff_success, buff_msg = self._apply_buff(character, tags, params, potency, duration_mult)
            if buff_success:
                success = True
                messages.append(buff_msg)

        # Process resistance effects
        if 'resistance' in tags:
            resist_success, resist_msg = self._apply_resistance(character, tags, params, potency, duration_mult)
            if resist_success:
                success = True
                messages.append(resist_msg)

        # Process utility effects (oils, polishes)
        if 'utility' in tags:
            utility_success, utility_msg = self._apply_utility(character, tags, params, potency, duration_mult)
            if utility_success:
                success = True
                messages.append(utility_msg)

        # Combine all messages
        final_message = " | ".join(messages) if messages else "No effect"

        # Add potency indicator if enhanced
        if success and potency > 1.0:
            final_message += f" (potency: {int(potency*100)}%)"

        return success, final_message

    def _apply_healing(
        self,
        character: 'Character',
        tags: list,
        params: dict,
        potency: float,
        duration_mult: float
    ) -> Tuple[bool, str]:
        """Apply healing effects (instant or over time)"""
        if 'instant' in tags:
            # Instant heal
            base_heal = params.get('heal_amount', 50)
            heal_amount = min(int(base_heal * potency), character.max_health - character.health)
            character.health += heal_amount
            return True, f"Restored {heal_amount:.0f} HP"

        elif 'over_time' in tags:
            # Heal over time (regeneration)
            base_regen = params.get('heal_per_second', 5.0)
            base_duration = params.get('duration', 60.0)

            actual_regen = base_regen * potency
            actual_duration = base_duration * duration_mult

            # Add regeneration buff
            buff = ActiveBuff(
                buff_id="potion_health_regen",
                name="Health Regeneration",
                effect_type="regenerate",
                category="health",
                magnitude="moderate",
                bonus_value=actual_regen,
                duration=actual_duration,
                duration_remaining=actual_duration,
                source="potion",
                consume_on_use=False
            )
            character.buffs.add_buff(buff)
            return True, f"Regenerating {actual_regen:.1f} HP/s for {actual_duration:.0f}s"

        return False, "Unknown healing type"

    def _apply_mana_restore(
        self,
        character: 'Character',
        tags: list,
        params: dict,
        potency: float,
        duration_mult: float
    ) -> Tuple[bool, str]:
        """Apply mana restoration effects (instant or over time)"""
        if 'instant' in tags:
            # Instant mana restore
            base_mana = params.get('mana_amount', 50)
            mana_amount = min(int(base_mana * potency), character.max_mana - character.mana)
            character.mana += mana_amount
            return True, f"Restored {mana_amount:.0f} Mana"

        elif 'over_time' in tags:
            # Mana over time (regeneration)
            base_regen = params.get('mana_per_second', 2.0)
            base_duration = params.get('duration', 60.0)

            actual_regen = base_regen * potency
            actual_duration = base_duration * duration_mult

            # Add mana regeneration buff
            buff = ActiveBuff(
                buff_id="potion_mana_regen",
                name="Mana Regeneration",
                effect_type="regenerate",
                category="mana",
                magnitude="moderate",
                bonus_value=actual_regen,
                duration=actual_duration,
                duration_remaining=actual_duration,
                source="potion",
                consume_on_use=False
            )
            character.buffs.add_buff(buff)
            return True, f"Regenerating {actual_regen:.1f} Mana/s for {actual_duration:.0f}s"

        return False, "Unknown mana restore type"

    def _apply_buff(
        self,
        character: 'Character',
        tags: list,
        params: dict,
        potency: float,
        duration_mult: float
    ) -> Tuple[bool, str]:
        """Apply stat buff effects"""
        buff_type = params.get('buff_type', 'strength')  # strength, defense, speed, max_hp, etc.
        base_value = params.get('buff_value', 0.2)  # Multiplier or flat value
        base_duration = params.get('duration', 300.0)

        actual_value = base_value * potency
        actual_duration = base_duration * duration_mult

        # Map buff type to category
        category_map = {
            'strength': 'combat',
            'defense': 'defense',
            'speed': 'movement',
            'max_hp': 'health'
        }
        category = category_map.get(buff_type, 'combat')

        # Map buff type to display name
        name_map = {
            'strength': 'Strength Boost',
            'defense': 'Defense Boost',
            'speed': 'Speed Boost',
            'max_hp': 'Vitality Boost'
        }
        buff_name = name_map.get(buff_type, f'{buff_type.capitalize()} Buff')

        # Create and add buff
        buff = ActiveBuff(
            buff_id=f"potion_{buff_type}",
            name=buff_name,
            effect_type="empower",
            category=category,
            magnitude="moderate",
            bonus_value=actual_value,
            duration=actual_duration,
            duration_remaining=actual_duration,
            source="potion",
            consume_on_use=False
        )
        character.buffs.add_buff(buff)

        # Format message based on buff type
        if buff_type == 'strength':
            return True, f"+{int(actual_value*100)}% physical damage for {actual_duration:.0f}s"
        elif buff_type == 'defense':
            return True, f"+{int(actual_value*100)}% defense for {actual_duration:.0f}s"
        elif buff_type == 'speed':
            return True, f"+{int(actual_value*100)}% speed for {actual_duration:.0f}s"
        elif buff_type == 'max_hp':
            return True, f"+{int(actual_value*100)}% max HP for {actual_duration:.0f}s"
        else:
            return True, f"{buff_type} buff for {actual_duration:.0f}s"

    def _apply_resistance(
        self,
        character: 'Character',
        tags: list,
        params: dict,
        potency: float,
        duration_mult: float
    ) -> Tuple[bool, str]:
        """Apply elemental resistance effects"""
        resistance_type = params.get('resistance_type', 'fire')  # fire, ice, elemental (all)
        base_reduction = params.get('damage_reduction', 0.5)  # 50% reduction
        base_duration = params.get('duration', 360.0)

        actual_reduction = min(base_reduction * potency, 0.9)  # Cap at 90%
        actual_duration = base_duration * duration_mult

        # Map resistance type to display name
        name_map = {
            'fire': 'Fire Resistance',
            'ice': 'Ice Resistance',
            'elemental': 'Elemental Resistance'
        }
        buff_name = name_map.get(resistance_type, f'{resistance_type.capitalize()} Resistance')

        # Create and add resistance buff
        buff = ActiveBuff(
            buff_id=f"resistance_{resistance_type}",
            name=buff_name,
            effect_type="fortify",
            category="defense",
            magnitude="moderate",
            bonus_value=actual_reduction,
            duration=actual_duration,
            duration_remaining=actual_duration,
            source="potion",
            consume_on_use=False
        )
        character.buffs.add_buff(buff)

        if resistance_type == 'elemental':
            return True, f"All elemental resistance for {actual_duration:.0f}s"
        else:
            return True, f"{int(actual_reduction*100)}% {resistance_type} resistance for {actual_duration:.0f}s"

    def _apply_utility(
        self,
        character: 'Character',
        tags: list,
        params: dict,
        potency: float,
        duration_mult: float
    ) -> Tuple[bool, str]:
        """Apply utility effects (efficiency oil, armor polish, weapon oil)"""
        utility_type = params.get('utility_type', 'efficiency')  # efficiency, armor, weapon
        base_value = params.get('utility_value', 0.15)
        base_duration = params.get('duration', 3600.0)

        actual_value = base_value * potency
        actual_duration = base_duration * duration_mult

        # Map utility type to category
        category_map = {
            'efficiency': 'gathering',
            'armor': 'defense',
            'weapon': 'combat'
        }
        category = category_map.get(utility_type, 'gathering')

        # Map utility type to display name
        name_map = {
            'efficiency': 'Efficiency Oil',
            'armor': 'Armor Polish',
            'weapon': 'Weapon Oil'
        }
        buff_name = name_map.get(utility_type, f'{utility_type.capitalize()} Enhancement')

        # Create and add utility buff
        buff = ActiveBuff(
            buff_id=f"utility_{utility_type}",
            name=buff_name,
            effect_type="empower",
            category=category,
            magnitude="moderate",
            bonus_value=actual_value,
            duration=actual_duration,
            duration_remaining=actual_duration,
            source="potion",
            consume_on_use=False
        )
        character.buffs.add_buff(buff)

        if utility_type == 'efficiency':
            return True, f"+{int(actual_value*100)}% gathering speed for {actual_duration:.0f}s"
        elif utility_type == 'armor':
            return True, f"+{int(actual_value*100)}% armor defense for {actual_duration:.0f}s"
        elif utility_type == 'weapon':
            return True, f"+{int(actual_value*100)}% weapon damage for {actual_duration:.0f}s"
        else:
            return True, f"{utility_type} enhancement for {actual_duration:.0f}s"


# Singleton instance
_potion_executor_instance = None

def get_potion_executor() -> PotionEffectExecutor:
    """Get singleton potion executor instance"""
    global _potion_executor_instance
    if _potion_executor_instance is None:
        _potion_executor_instance = PotionEffectExecutor()
    return _potion_executor_instance

File: Unity/Assets/Scripts/Game1.Systems/Items/PotionSystem.cs
// ============================================================================
// Game1.Systems.Items.PotionSystem
// Migrated from: systems/potion_system.py (387 lines)
// Migration phase: 4
// Date: 2026-02-13
//
// Tag-driven potion effect application system.
// Supported tags: healing, mana_restore, buff, resistance, utility.
// Tag modifiers: instant, over_time, self.
// Quality multipliers from crafting: potency (effect strength), duration.
// ============================================================================

using System;
using System.Collections.Generic;

namespace Game1.Systems.Items
{
    // ========================================================================
    // Potion Character Interface
    // ========================================================================

    /// <summary>
    /// Interface for character state needed by the potion system.
    /// Decouples PotionSystem from concrete Character class.
    /// </summary>
    public interface IPotionTarget
    {
        float Health { get; set; }
        float MaxHealth { get; }
        float Mana { get; set; }
        float MaxMana { get; }

        /// <summary>
        /// Add a timed buff to the character.
        /// </summary>
        void AddBuff(PotionBuff buff);
    }

    // ========================================================================
    // Potion Buff
    // ========================================================================

    /// <summary>
    /// A timed buff applied by a potion. Matches Python ActiveBuff used by potions.
    /// </summary>
    public class PotionBuff
    {
        public string BuffId { get; set; }
        public string Name { get; set; }
        public string EffectType { get; set; }     // "regenerate", "empower", "fortify"
        public string Category { get; set; }        // "health", "mana", "combat", "defense", etc.
        public string Magnitude { get; set; } = "moderate";
        public float BonusValue { get; set; }
        public float Duration { get; set; }
        public float DurationRemaining { get; set; }
        public string Source { get; set; } = "potion";
        public bool ConsumeOnUse { get; set; }
    }

    // ========================================================================
    // Potion Item Definition (subset used by this system)
    // ========================================================================

    /// <summary>
    /// Potion-relevant fields extracted from a MaterialDefinition.
    /// </summary>
    public class PotionDefinition
    {
        public string Name { get; set; }
        public List<string> EffectTags { get; set; } = new();
        public object EffectParams { get; set; } // Dictionary or List of Dictionaries
    }

    // ========================================================================
    // Potion System
    // ========================================================================

    /// <summary>
    /// Executes potion effects based on tags and parameters.
    /// Singleton pattern matches Python get_potion_executor().
    ///
    /// Supported effect tags (from Python PotionEffectExecutor):
    ///   healing + instant: Restore HP immediately
    ///   healing + over_time: HP regeneration buff
    ///   mana_restore + instant: Restore mana immediately
    ///   mana_restore + over_time: Mana regeneration buff
    ///   buff: Apply stat buff (strength, defense, speed, max_hp, attack_speed)
    ///   resistance: Elemental damage resistance (fire, ice, elemental)
    ///   utility: Tool/armor enhancements (efficiency, armor, weapon)
    /// </summary>
    public class PotionSystem
    {
        private static PotionSystem _instance;

        public static PotionSystem Instance => _instance ??= new PotionSystem();

        // ====================================================================
        // Main Entry Point
        // ====================================================================

        /// <summary>
        /// Apply potion effects to character based on tags and parameters.
        /// Returns (success, message).
        /// Matches Python apply_potion_effect() exactly.
        /// </summary>
        public (bool success, string message) UsePotion(
            IPotionTarget character,
            PotionDefinition potionDef,
            Dictionary<string, object> craftedStats = null)
        {
            if (potionDef.EffectTags == null || potionDef.EffectTags.Count == 0)
                return (false, $"{potionDef.Name} has no effect tags defined");

            // Get quality multipliers from crafting stats
            float potency = 1.0f;
            float durationMult = 1.0f;

            if (craftedStats != null)
            {
                if (craftedStats.TryGetValue("potency", out var potObj))
                    potency = Convert.ToSingle(potObj) / 100f;
                if (craftedStats.TryGetValue("duration", out var durObj))
                    durationMult = Convert.ToSingle(durObj) / 100f;
            }

            var tags = potionDef.EffectTags;
            var messages = new List<string>();
            bool success = false;

            // Support modular effectParams (array or single dict)
            if (potionDef.EffectParams is IList<object> paramsList)
            {
                foreach (var paramObj in paramsList)
                {
                    if (paramObj is Dictionary<string, object> paramSet)
                    {
                        var (s, m) = ApplySingleEffect(character, tags, paramSet, potency, durationMult);
                        if (s) { success = true; messages.Add(m); }
                    }
                }
            }
            else if (potionDef.EffectParams is Dictionary<string, object> singleParams)
            {
                var (s, m) = ApplySingleEffect(character, tags, singleParams, potency, durationMult);
                if (s) { success = true; messages.Add(m); }
            }

            string finalMessage = messages.Count > 0 ? string.Join(" | ", messages) : "No effect";

            if (success && potency > 1.0f)
                finalMessage += $" (potency: {(int)(potency * 100)}%)";

            return (success, finalMessage);
        }

        // ====================================================================
        // Single Effect Application
        // ====================================================================

        private (bool, string) ApplySingleEffect(
            IPotionTarget character,
            List<string> tags,
            Dictionary<string, object> parms,
            float potency,
            float durationMult)
        {
            if (tags.Contains("healing"))
                return ApplyHealing(character, tags, parms, potency, durationMult);
            if (tags.Contains("mana_restore"))
                return ApplyManaRestore(character, tags, parms, potency, durationMult);
            if (tags.Contains("buff"))
                return ApplyBuff(character, tags, parms, potency, durationMult);
            if (tags.Contains("resistance"))
                return ApplyResistance(character, tags, parms, potency, durationMult);
            if (tags.Contains("utility"))
                return ApplyUtility(character, tags, parms, potency, durationMult);

            return (false, "Unknown effect type");
        }

        // ====================================================================
        // Healing (instant or over_time)
        // ====================================================================

        private (bool, string) ApplyHealing(
            IPotionTarget character, List<string> tags,
            Dictionary<string, object> parms, float potency, float durationMult)
        {
            if (tags.Contains("instant"))
            {
                float baseHeal = GetFloat(parms, "heal_amount", 50f);
                float healAmount = MathF.Min(baseHeal * potency, character.MaxHealth - character.Health);
                character.Health += healAmount;
                return (true, $"Restored {healAmount:F0} HP");
            }

            if (tags.Contains("over_time"))
            {
                float baseRegen = GetFloat(parms, "heal_per_second", 5f);
                float baseDuration = GetFloat(parms, "duration", 60f);
                float actualRegen = baseRegen * potency;
                float actualDuration = baseDuration * durationMult;

                character.AddBuff(new PotionBuff
                {
                    BuffId = "potion_health_regen",
                    Name = "Health Regeneration",
                    EffectType = "regenerate",
                    Category = "health",
                    BonusValue = actualRegen,
                    Duration = actualDuration,
                    DurationRemaining = actualDuration,
                });
                return (true, $"Regenerating {actualRegen:F1} HP/s for {actualDuration:F0}s");
            }

            return (false, "Unknown healing type");
        }

        // ====================================================================
        // Mana Restore (instant or over_time)
        // ====================================================================

        private (bool, string) ApplyManaRestore(
            IPotionTarget character, List<string> tags,
            Dictionary<string, object> parms, float potency, float durationMult)
        {
            if (tags.Contains("instant"))
            {
                float baseMana = GetFloat(parms, "mana_amount", 50f);
                float manaAmount = MathF.Min(baseMana * potency, character.MaxMana - character.Mana);
                character.Mana += manaAmount;
                return (true, $"Restored {manaAmount:F0} Mana");
            }

            if (tags.Contains("over_time"))
            {
                float baseRegen = GetFloat(parms, "mana_per_second", 2f);
                float baseDuration = GetFloat(parms, "duration", 60f);
                float actualRegen = baseRegen * potency;
                float actualDuration = baseDuration * durationMult;

                character.AddBuff(new PotionBuff
                {
                    BuffId = "potion_mana_regen",
                    Name = "Mana Regeneration",
                    EffectType = "regenerate",
                    Category = "mana",
                    BonusValue = actualRegen,
                    Duration = actualDuration,
                    DurationRemaining = actualDuration,
                });
                return (true, $"Regenerating {actualRegen:F1} Mana/s for {actualDuration:F0}s");
            }

            return (false, "Unknown mana restore type");
        }

        // ====================================================================
        // Buff (stat buffs: strength, defense, speed, max_hp, attack_speed)
        // ====================================================================

        private (bool, string) ApplyBuff(
            IPotionTarget character, List<string> tags,
            Dictionary<string, object> parms, float potency, float durationMult)
        {
            string buffType = GetString(parms, "buff_type", "strength");
            float baseValue = GetFloat(parms, "buff_value", 0.2f);
            float baseDuration = GetFloat(parms, "duration", 300f);

            float actualValue = baseValue * potency;
            float actualDuration = baseDuration * durationMult;

            var categoryMap = new Dictionary<string, string>
            {
                ["strength"] = "combat",
                ["defense"] = "defense",
                ["speed"] = "movement",
                ["max_hp"] = "health",
                ["attack_speed"] = "combat",
            };
            string category = categoryMap.TryGetValue(buffType, out var c) ? c : "combat";

            var nameMap = new Dictionary<string, string>
            {
                ["strength"] = "Strength Boost",
                ["defense"] = "Defense Boost",
                ["speed"] = "Speed Boost",
                ["max_hp"] = "Vitality Boost",
                ["attack_speed"] = "Attack Speed Boost",
            };
            string buffName = nameMap.TryGetValue(buffType, out var n) ? n : $"{buffType} Buff";

            character.AddBuff(new PotionBuff
            {
                BuffId = $"potion_{buffType}",
                Name = buffName,
                EffectType = "empower",
                Category = category,
                BonusValue = actualValue,
                Duration = actualDuration,
                DurationRemaining = actualDuration,
            });

            string msg = buffType switch
            {
                "strength" => $"+{(int)(actualValue * 100)}% physical damage for {actualDuration:F0}s",
                "defense" => $"+{(int)(actualValue * 100)}% defense for {actualDuration:F0}s",
                "speed" => $"+{(int)(actualValue * 100)}% move speed for {actualDuration:F0}s",
                "attack_speed" => $"+{(int)(actualValue * 100)}% attack speed for {actualDuration:F0}s",
                "max_hp" => $"+{(int)(actualValue * 100)}% max HP for {actualDuration:F0}s",
                _ => $"{buffType} buff for {actualDuration:F0}s",
            };

            return (true, msg);
        }

        // ====================================================================
        // Resistance (fire, ice, elemental)
        // ====================================================================

        private (bool, string) ApplyResistance(
            IPotionTarget character, List<string> tags,
            Dictionary<string, object> parms, float potency, float durationMult)
        {
            string resistType = GetString(parms, "resistance_type", "fire");
            float baseReduction = GetFloat(parms, "damage_reduction", 0.5f);
            float baseDuration = GetFloat(parms, "duration", 360f);

            float actualReduction = MathF.Min(baseReduction * potency, 0.9f); // Cap at 90%
            float actualDuration = baseDuration * durationMult;

            var nameMap = new Dictionary<string, string>
            {
                ["fire"] = "Fire Resistance",
                ["ice"] = "Ice Resistance",
                ["elemental"] = "Elemental Resistance",
            };
            string buffName = nameMap.TryGetValue(resistType, out var n) ? n : $"{resistType} Resistance";

            character.AddBuff(new PotionBuff
            {
                BuffId = $"resistance_{resistType}",
                Name = buffName,
                EffectType = "fortify",
                Category = "defense",
                BonusValue = actualReduction,
                Duration = actualDuration,
                DurationRemaining = actualDuration,
            });

            string msg = resistType == "elemental"
                ? $"All elemental resistance for {actualDuration:F0}s"
                : $"{(int)(actualReduction * 100)}% {resistType} resistance for {actualDuration:F0}s";

            return (true, msg);
        }

        // ====================================================================
        // Utility (efficiency oil, armor polish, weapon oil)
        // ====================================================================

        private (bool, string) ApplyUtility(
            IPotionTarget character, List<string> tags,
            Dictionary<string, object> parms, float potency, float durationMult)
        {
            string utilType = GetString(parms, "utility_type", "efficiency");
            float baseValue = GetFloat(parms, "utility_value", 0.15f);
            float baseDuration = GetFloat(parms, "duration", 3600f);

            float actualValue = baseValue * potency;
            float actualDuration = baseDuration * durationMult;

            var categoryMap = new Dictionary<string, string>
            {
                ["efficiency"] = "gathering",
                ["armor"] = "defense",
                ["weapon"] = "combat",
            };
            string category = categoryMap.TryGetValue(utilType, out var c) ? c : "gathering";

            var nameMap = new Dictionary<string, string>
            {
                ["efficiency"] = "Efficiency Oil",
                ["armor"] = "Armor Polish",
                ["weapon"] = "Weapon Oil",
            };
            string buffName = nameMap.TryGetValue(utilType, out var n) ? n : $"{utilType} Enhancement";

            character.AddBuff(new PotionBuff
            {
                BuffId = $"utility_{utilType}",
                Name = buffName,
                EffectType = "empower",
                Category = category,
                BonusValue = actualValue,
                Duration = actualDuration,
                DurationRemaining = actualDuration,
            });

            string msg = utilType switch
            {
                "efficiency" => $"+{(int)(actualValue * 100)}% gathering speed for {actualDuration:F0}s",
                "armor" => $"+{(int)(actualValue * 100)}% armor defense for {actualDuration:F0}s",
                "weapon" => $"+{(int)(actualValue * 100)}% weapon damage for {actualDuration:F0}s",
                _ => $"{utilType} enhancement for {actualDuration:F0}s",
            };

            return (true, msg);
        }

        // ====================================================================
        // Param Helpers
        // ====================================================================

        private static float GetFloat(Dictionary<string, object> d, string key, float defaultVal)
        {
            return d.TryGetValue(key, out var v) ? Convert.ToSingle(v) : defaultVal;
        }

        private static string GetString(Dictionary<string, object> d, string key, string defaultVal)
        {
            return d.TryGetValue(key, out var v) ? v?.ToString() ?? defaultVal : defaultVal;
        }
    }
}
